### ISSUE TITLE
conch gmpy support doesn't work anymore

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2924]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#2924|
|Type|defect|
|Created|2007-11-23 19:28:13Z|
|Branch|https://github.com/twisted/twisted/tree/conch-gmpy-support-2924|

This fails with this error:
```
  File "twisted-trunk/twisted/conch/ssh/common.py", line 105, in _fastgetMP
    mp.append(long(gmpy.mpz(data[c+4:c+l+4][::-1]+'\x00')))
exceptions.ValueError: string without NULL characters expected
```

<details><summary>Searchable metadata</summary>

```
trac-id__2924 2924
type__defect defect
reporter__therve therve
priority__highest highest
milestone__ 
branch__branches_conch_gmpy_support_2924 branches/conch-gmpy-support-2924
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1195846093000000 1195846093000000
changetime__1198343469000000 1198343469000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

It comes from [21670]... It's not the first problem spotted since this merge, I wonder if it shouldn't be reverted.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

One step in the right direction would be to get gmp onto at least one of the slaves.  Then, there should be unit tests which run both versions of the code, even if gmp is installed.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [21979]) Branching to 'conch-gmpy-support-2924'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve removed owner|
|-|-|

OK this is ready to review. Although it's useless until a slave got gmpy installed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

bot-cube has gmpy available now

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve<br>@exarkun set status to `new`|
|-|-|

The indentation in MPTestCase.test_multipleGetMP is a bit funny.  The 2nd line of the string literal should be indented two spaces fewer than it is, I think.

It might be nice to have a test for an integer > 255.

Otherwise, nice fix, thanks.

As far as reverting the original branch goes, I think we probably missed that opportunity. :(  I'm guessing that untangling what has happened since will be more work than just fixing the problems we find.  Sorry for not being more rigorous on that review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

(In [22121]) Fix indentation, add one test.

Refs [#2924](https://github.com/twisted/twisted/issues/2924)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve removed owner|
|-|-|

Thanks for the review, I've made the modifications.

Regarding the revert, the branch was big enough so it wasn't very easy to make it land. Sometimes we just have to go forward :).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @therve|
|-|-|

Looks good, please merge

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve commented|
|-|-|

(In [22122]) Skip the tests when pyCrypto is not there.

Refs [#2924](https://github.com/twisted/twisted/issues/2924)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set status to `closed`|
|-|-|

(In [22123]) Merge conch-gmpy-support-2924

Author: therve
Reviewer: exarkun
Fixes [#2924](https://github.com/twisted/twisted/issues/2924)

Make gmpy support in conch works again, and add tests for both standard and fast
implementation of multiple precision. It was broken in r21670.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
